src/StmtToViz.cpp

```diff
@@ -0,0 +1,2900 @@
+            print_html_element("span", "matched", "]");
+            if (i < op->bounds.size() - 1) {
+                print_html_element("span", "matched", ", ");
+            }
+        }
+        print_html_element("span", "matched", ")");
+
+        // Print condition
+        if (!is_const_one(op->condition)) {
+            print_html_element("span", "keyword", " if ");
+            print(op->condition);
+        }
+
+        // Print prefetch body
+        print_opening_tag("div", "indent PrefetchBody");
+        print(op->body);
+        print_closing_tag("div");
+
+        print_closing_tag("div");
+    }
+
+    void visit(const Atomic *op) override {
+        // Give this node a unique id
+        int id = gen_unique_id();
+
+        // Start a dive to hold code for this atomic
+        print_opening_tag("div", "Atomic");
+
+        // Generate the show hide icon/text buttons
+        print_toggle_anchor_opening_tag(id);
+
+        // -- print icon
+        print_show_hide_icon(id);
+
+        // -- print text
+        print_html_element("span", "matched keyword", "atomic");
+        if (!op->mutex_name.empty()) {
+            print_html_element("span", "matched", "(");
+            print_html_element("span", "Symbol", op->mutex_name);
+            print_html_element("span", "matched", ")");
+        }
+
+        print_toggle_anchor_closing_tag();
+
+        // Open code block to hold atomic body
+        print_html_element("span", "matched", " {");
+
+        // Open indented div to hold atomic code
+        print_opening_tag("div", "indent AtomicBody", id);
+
+        // Print atomic body
+        print(op->body);
+
+        // Close indented div holding body code
+        print_closing_tag("div");
+
+        // Close code block holding fork body
+        print_html_element("span", "matched ClosingBrace cb-" + std::to_string(id), "}");
+
+        // Close div holding this atomic
+        print_closing_tag("div");
+    }
+};
+
+/************** HTMLVisualizationPrinter Class ***************/
+// Visualizes the IR in HTML. The visualization is essentially
+// an abstracted version of the code, highlighting the higher
+// level execution pipeline along with key properties of the
+// execution performed at each stage.
+class HTMLVisualizationPrinter : public IRVisitor {
+public:
+    HTMLVisualizationPrinter(std::ofstream &os)
+        : stream(os), printer(ss), _id(0) {
+    }
+
+    void init_cost_info(IRCostModel cost_model) {
+        _cost_model = std::move(cost_model);
+    }
+
+    void print(const Module &m, AssemblyInfo asm_info) {
+        _assembly_info = std::move(asm_info);
+        for (const auto &fn : m.functions()) {
+            print(fn);
+        }
+    }
+
+private:
+    // Handle to output file stream
+    std::ofstream &stream;
+
+    // Used to track the context within generated HTML
+    std::vector<string> context_stack_tags;
+
+    // Used to translate IR to code in HTML
+    ostringstream ss;
```

Comment:
Is this ostringsteam used in a way that is accumulates between function calls? At first glance I don't see that, in which case it seems simpler and more robust to just declare it locally where needed, but maybe I'm missing something...